### PR TITLE
Add deployment pipeline

### PR DESCRIPTION
**Pull request by liampetch**

### CODE REVIEW

File: Deploymentpipe [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+            sh 'printenv'
+            checkout scm
+            sh 'git clean -fdx'
+        }
+
+        def HASH = sh(returnStdout: true, script: 'git rev-parse --short HEAD').trim()
+        def DOCKER_TAG = "${params.DOCKER_REPO}:${HASH}"
+
+        stage ('Build base') {
+           sh 'make docker-base'
```
Comment (julesjcraske) [inline Deploymentpipe]: You don't want to rebuild the base image every run

File: Deploymentpipe [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        }
+
+        stage ('Deps for deploy') {
+            sh 'cp .env.jenkins .env'
```
Comment (julesjcraske) [inline Deploymentpipe]: We don't need an env file for jenkins

File: Deploymentpipe [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+            sh 'echo "LOCAL_USER_ID=$(id -u)" >> .env'
+            sh 'docker-compose down --remove-orphans'
+            sh 'docker-compose build'
+            withCredentials([string(credentialsId: 'github-oauth-token', variable: 'GITHUB_OAUTH_TOKEN')]) {
```
Comment (julesjcraske) [inline Deploymentpipe]: We don't use any private repositories from our github account as dependencies, you don't need to do this.

File: Deploymentpipe [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+        stage ('Build base') {
+           sh 'make docker-base'
+        }
+
+        stage ('Deps for deploy') {
+            sh 'cp .env.jenkins .env'
+            sh 'echo "LOCAL_USER_ID=$(id -u)" >> .env'
+            sh 'docker-compose down --remove-orphans'
+            sh 'docker-compose build'
```
Comment (julesjcraske) [inline Deploymentpipe]: We don't have anything to build

File: Deploymentpipe [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        }
+
+        stage ('Push image') {
+            sh "aws ecr get-login --no-include-email --region ${params.REGION} | /bin/bash"
+            sh "docker push ${DOCKER_TAG}"
+        }
+
+        stage ('Terraform') {
+            sh """
+                ./terraform.sh init -input=false -backend-config backends/${params.TERRAFORM_BACKEND}.tfvars
```
Comment (julesjcraske) [inline Deploymentpipe]: You might want to use a shell script like ./bin/deploy in p4l-main which make it easier to run these deployment scripts yourself locally

File: Dockerpipe [hunk 6]
```diff
@@ -30,3 +30,16 @@ node { ansiColor('xterm') { timestamps {
         }
     }
 } } }
+
+if (env.BRANCH_NAME ==~ /^(master|add-deploy-pipeline)$/) {
+    stage ('Deploy to Production') {
+        build job: "Deploymentpipe/phpwkhtml-server/$BRANCH_NAME",
```
Comment (julesjcraske) [inline Dockerpipe]: This job name is incorrect

File: Dockerpipe [hunk 7]
```diff
@@ -30,3 +30,16 @@ node { ansiColor('xterm') { timestamps {
         }
     }
 } } }
+
+if (env.BRANCH_NAME ==~ /^(master|add-deploy-pipeline)$/) {
```
Comment (julesjcraske) [inline Dockerpipe]: Remove the specific name, test this by editing the script on jenkins using replay

**Comment by mclueppers** [comment]
With every commit you do it's getting better and better. But we should stop at some point and just run it. Improvements as we go is what we are supposed to do else we might over-engineer it :)

**Pull request merged**